pkg/app/launcher/launcher.go

```diff
@@ -0,0 +1,369 @@
+			state.Status = AppStatusRunning
+		}
+		states = append(states, state)
+	}
+	return states
+}
+
+// StartApp starts cmd with given args and env.
+// If 'args' is nil, default args will be used.
+func (l *Launcher) StartApp(cmd string, args, envs []string) error {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	return l.startApp(cmd, args, envs)
+}
+
+func (l *Launcher) startApp(cmd string, args, envs []string) error {
+	log := l.log.WithField("func", "StartApp").WithField("cmd", cmd)
+
+	// Obtain associated app config.
+	ac, ok := l.apps[cmd]
+	if !ok {
+		return ErrAppNotFound
+	}
+	if args != nil {
+		ac.Args = args
+	}
+
+	// Make proc config.
+	procConf, err := makeProcConfig(l.conf, ac, envs)
+	if err != nil {
+		return err
+	}
+
+	// Start proc and persist pid.
+	pid, err := l.procM.Start(procConf)
+	if err != nil {
+		return err
+	}
+	if err := l.persistPID(cmd, pid); err != nil {
+		log.WithError(err).Warn("Failed to persist pid.")
+	}
+
+	return nil
+}
+
+// StopApp stops running app.
+func (l *Launcher) StopApp(name string) (*appserver.Proc, error) {
+	log := l.log.WithField("func", "StopApp").WithField("app_name", name)
+
+	proc, ok := l.procM.ProcByName(name)
+	if !ok {
+		return nil, ErrAppNotRunning
+	}
+
+	l.log.Info("Stopping app...")
+
+	if err := l.procM.Stop(name); err != nil {
+		log.WithError(err).Warn("Failed to stop app.")
+		return proc, err
+	}
+
+	return proc, nil
+}
+
+// RestartApp restarts a running app.
+func (l *Launcher) RestartApp(name string) error {
+	l.log.WithField("func", "RestartApp").WithField("app_name", name).
+		Info("Restarting app...")
+
+	proc, err := l.StopApp(name)
+	if err != nil {
+		return fmt.Errorf("failed to stop %s: %w", name, err)
+	}
+
+	cmd := proc.Cmd()
+	if err := l.StartApp(name, cmd.Args, cmd.Env); err != nil {
+		return fmt.Errorf("failed to start %s: %w", name, err)
+	}
+
+	return nil
+}
+
+func makeProcConfig(lc Config, ac AppConfig, envs []string) (appcommon.ProcConfig, error) {
+	procConf := appcommon.ProcConfig{
+		AppName:     ac.Name,
+		AppSrvAddr:  lc.ServerAddr,
+		ProcKey:     appcommon.RandProcKey(),
+		ProcArgs:    ac.Args,
+		ProcEnvs:    envs,
+		ProcWorkDir: filepath.Join(lc.LocalPath, ac.Name),
+		VisorPK:     lc.VisorPK,
+		RoutingPort: ac.Port,
+		BinaryLoc:   filepath.Join(lc.BinPath, ac.Name),
+		LogDBLoc:    filepath.Join(lc.LocalPath, ac.Name+"_log.db"),
+	}
+	err := ensureDir(&procConf.ProcWorkDir)
+	return procConf, err
+}
+
+func ensureDir(path *string) error {
+	var err error
+	if *path, err = filepath.Abs(*path); err != nil {
+		return fmt.Errorf("failed to expand path: %s", err)
+	}
+	if _, err := os.Stat(*path); !os.IsNotExist(err) {
+		return nil
+	}
+	if err := os.MkdirAll(*path, 0750); err != nil {
+		return fmt.Errorf("failed to create dir: %s", err)
+	}
+	return nil
+}
+
+/*
+	<<< PID management >>>
+*/
+
+func (l *Launcher) pidFile() (*os.File, error) {
+	return os.OpenFile(filepath.Join(l.conf.LocalPath, "apps-pid.txt"), os.O_RDWR|os.O_CREATE, 0600)
```

Comment:
I know it's probably an old part, but could you please move the file name to a constant?